### ISSUE TITLE
attack popup target changing when clicked

### ISSUE THREAD

**Issue by zdechlak** [forkserf/forkserf#325]
https://user-images.githubusercontent.com/2237225/218204637-d48141a8-abca-4332-baae-3e81b9c951f8.mp4

when I sent my army - the building icon changed - but the attack was going to the right building

**Comment by tlongstretch** [comment]
this is odd, I will try to reproduce this

**Comment by zdechlak** [comment]
also look at how these numerical values behave

**Comment by tlongstretch** [comment]
I never knew those knight-distance buttons were clickable, I thought they were static.  Looks like they are indeed supposed to only send the knights up to a certain distance

```
  case ACTION_ATTACKING_SELECT_ALL_1:
    player->knights_attacking = player->attacking_knights[0];
    break;
  case ACTION_ATTACKING_SELECT_ALL_2:
    player->knights_attacking = player->attacking_knights[0]
                                + player->attacking_knights[1];
    break;
  case ACTION_ATTACKING_SELECT_ALL_3:
    player->knights_attacking = player->attacking_knights[0]
                                + player->attacking_knights[1]
                                + player->attacking_knights[2];
    break;
  case ACTION_ATTACKING_SELECT_ALL_4:
    player->knights_attacking = player->attacking_knights[0]
                                + player->attacking_knights[1]
                                + player->attacking_knights[2]
                                + player->attacking_knights[3];
```

**Comment by tlongstretch** [comment]
I haven't been able to reproduce this. Is it happening reliably in a savegame you have?  Or is it happening often?  rarely?  only this once?

**Comment by zdechlak** [comment]
very rarely, i will try to recreate it
I was playing on a map with 4 AIs and jumping between them with a key on the keyboard

However, it seems to me that the problem also appeared in the game human vs computer

**Comment by zdechlak** [comment]
_I never knew those knight-distance buttons were clickable, I thought they were static. Looks like they are indeed supposed to only send the knights up to a certain distance_

it seems to me that it's about the strength of the attack - who to send to the front, depending on the option in the army panel

so, 1 icon is the strongest knight - and the last one is the weakest

this is how it works in https://www.simpleguide.net/sg/serflings.jsf

but I'm not sure how it was in the original game - and I'll play it on the Amiga in my spare time to understand it

see this is what i'm reading: https://www.oldgames.sk/downloads/oldgames/strategy/Settlers/Docs/manual-settlers-1993.pdf

**Comment by tlongstretch** [comment]
oh, it is possible that the AI is itself "opening the attack popup" in a hidden way when declaring its attacks which might cycle the target.  Let me know if you see this happen while playing as a human player.  Manipulating AI players is sometimes unpredictable though it mostly works, but I don't want to spend much time debugging issues caused by this.

**Comment by tlongstretch** [comment]
> _I never knew those knight-distance buttons were clickable, I thought they were static. Looks like they are indeed supposed to only send the knights up to a certain distance_
> 
> it seems to me that it's about the strength of the attack - who to send to the front, depending on the option in the army panel
> 
> so, 1 icon is the strongest knight - and the last one is the weakest
> 
> this is how it works in https://www.simpleguide.net/sg/serflings.jsf
> 
> but I'm not sure how it was in the original game - and I'll play it on the Amiga in my spare time to understand it
> 
> see this is what i'm reading: https://www.oldgames.sk/downloads/oldgames/strategy/Settlers/Docs/manual-settlers-1993.pdf


No, this is about the *distance* the knights are from the building being targeted for attack.  Notice that the pictures are of knights in various states of readiness, from very active to asleep.  Clicking those boxes says "send only knights who are this distance or closer as part of the attack".    I believe this is to prevent sending knights long distances to attack when sufficient nearby knights could be sent.

The strength of the attack is purely based on knight rank, gold-morale of attacking player, and the number of knights sent.  And the "send strong" vs "send weak" toggle button determines if the strongest available (at that distance) or weakest available sent
